src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -0,0 +1,159 @@
+                TimeSpan timeout,
+                ReferenceCountedDisposable<RemotableDataJsonRpc> remotableDataRpc)
+            {
+                _hubClient = hubClient;
+                _hostGroup = hostGroup;
+                _timeout = timeout;
+                _remotableDataRpc = remotableDataRpc;
+
+                _maxPoolConnections = maxPoolConnection;
+
+                // we have 4 services. so start from 4. later if we add more services, it will still work.
+                _pools = new ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>>(concurrencyLevel: 4, capacity: 4);
+
+                _usePool = usePool;
+                _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+            }
+
+            public Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                // pool is turned off either by option or pool has shutdown.
+                if (!_usePool)
+                {
+                    // RemoteHostClient is allowed to be restarted by IRemoteHostClientService.RequestNewRemoteHostAsync
+                    // when that happen, existing RemoteHostClient doesn't suddenly go away and start to throw exception. it handles
+                    // shutdown gracefully. it will keep serving existing requests or even new requests if someone is holding old remoteHostClient.
+                    // all connections made to old RemoteHost will eventually go away and that's when RemoteHost is actually removed.
+                    // simplyput, RequestNewRemoteHostAsync is not intrusive to running features. all existing one will keep do what is doing
+                    // with old RemoteHost and only new request will go to new remoteHost.
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                // when callbackTarget is given, we can't share/pool connection since callbackTarget attaches a state to connection.
+                // so connection is only valid for that specific callbackTarget. it is up to the caller to keep connection open
+                // if he wants to reuse same connection
+                if (callbackTarget != null)
+                {
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                return TryGetConnectionFromPoolAsync(serviceName, cancellationToken);
+            }
+
+            private async Task<Connection> TryGetConnectionFromPoolAsync(string serviceName, CancellationToken cancellationToken)
+            {
+                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<JsonRpcConnection>());
+                if (queue.TryDequeue(out var connection))
+                {
+                    return new PooledConnection(this, serviceName, connection);
```

Comment:
in normal case, due to connection pool, I dont think it will matter much. let's see whether that type shows up in trace at all. I think it won't and normally won't be created more than 20 per solution for entire VS session.

Reply:
ah, yes. JsonRpcConnection is the one that actually pooled you are right :). but PooledConnection doesn't get created more than what JsonRpcConnection has been created. so, I dont think there is added impact on allocation here.